src/bookmarks.py

```diff
@@ -0,0 +1,84 @@
+# devices.py
+#
+# Copyright 2025 Jason Beetham
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from gi.repository import Gio, GLib, GObject
+
+from . import logger
+from . import utils
+import json, os
+
+
+class PortfolioBookmarks(GObject.GObject):
+    __gtype_name__ = "PortfolioBookmarks"
+
+    __gsignals__ = {
+        "toggle-bookmark": (GObject.SignalFlags.RUN_LAST, None, (str,)),
```

Comment:
I don't think you need this "toggle" signal, in any case should be a public method, to be consumed by UI, which in turn will trigger add or remove signals for the UI to do what's needs to be done (e.g., remove or add a new place under bookmarks).